### PR TITLE
chore: upgrade to Node.JS v18 LTS

### PR DESCRIPTION
**Pull request by pshenmic**
## Issue being fixed or feature implemented

The project still using node v16, while the Node.JS LTS have already upgraded to v18 a long time ago. There is no more obstacles to migrate and project successfully builds and runs on the Node v18. 

## What was done?

* Updated documentation
* Upgraded Node.JS Github Action runners
* Upgraded basebuild image in dapi, dashmate, js-drive and platform-test-suite
* Upgraded wasm-re2 module

## How Has This Been Tested?

Locally + CI

## Breaking Changes

* Minimum supported Node.JS version was increased to v18
* Upgraded wasm-re2 module to v2.0.4 (not compatible with Node v16 anymore)
* Removed `node-fetch` from dashmate

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by antouhou — DISMISSED** [review]
👍

**Comment by strophy** [comment]
When i run `yarn build` it generates some new proto files, i think these changes should be included here?
```
        modified:   packages/dapi-grpc/clients/platform/v0/nodejs/platform_pbjs.js
        modified:   packages/dapi-grpc/clients/platform/v0/nodejs/platform_protoc.js
        modified:   packages/dapi-grpc/clients/platform/v0/objective-c/Platform.pbobjc.h
        modified:   packages/dapi-grpc/clients/platform/v0/objective-c/Platform.pbobjc.m
        modified:   packages/dapi-grpc/clients/platform/v0/python/platform_pb2.py
        modified:   packages/dapi-grpc/clients/platform/v0/web/platform_pb.d.ts
        modified:   packages/dapi-grpc/clients/platform/v0/web/platform_pb.js
```

File: packages/dashmate/.eslintrc [hunk 1]
```diff
@@ -4,6 +4,9 @@
     // Required for certain syntax usages
     "ecmaVersion": 2022
   },
+  "globals": {
```
Comment (shumkov) [inline packages/dashmate/.eslintrc:7]: I think you just need to disable browser env

File: packages/js-dapi-client/polyfills/fetch-polyfill.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 const { default: fetch, Headers, Request, Response } = require('node-fetch');
```
Comment (shumkov) [inline packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: Is it polyfill for browser?

File: packages/dashmate/.eslintrc [hunk 3]
```diff
@@ -4,6 +4,9 @@
...
+    "fetch": false
```
Comment (shumkov) [inline packages/dashmate/.eslintrc:8]: Why this is set to false?

**Review by shumkov — CHANGES_REQUESTED** [review]
You also need to update Docker image as well

File: packages/js-dapi-client/polyfills/fetch-polyfill.js (hunk 2 above)
Comment (pshenmic) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: No, it is polyfill for Node.js, because fetch already bundled in both environments, but in Node environment it is based on undici, that is hard to bundle in the web yet., so it is being replace with node fetch

File: packages/dashmate/.eslintrc (hunk 3 above)
Comment (pshenmic) [reply packages/dashmate/.eslintrc:8]: Because fetch is now bundled with Node.js, and eslint errors about undefined function being used. The smarter way would be upgrade babel, eslint and the rest of the libraries to the recent version. I have already created a task for that

Comment (pshenmic) [reply packages/dashmate/.eslintrc:8]: https://github.com/dashpay/platform/issues/1296

**Comment by pshenmic** [comment]
@strophy this is something forgotten to be added to the .gitignore in the v25 branch I believed. This pr does not change this behaviour

@shumkov What docker images did I forget to add?

File: packages/dashmate/.eslintrc (hunk 3 above)
Comment (shumkov) [reply packages/dashmate/.eslintrc:8]: I see thank you

**Comment by strophy** [comment]
@pshenmic I guess here: https://github.com/dashpay/platform/blob/feat/node-lts-v25/Dockerfile#L56
Probably need to update Alpine to get Node 18, you had this update prepared somewhere else I think?

**Comment by strophy** [comment]
@pshenmic please review the entire Dockerfile, there are other sections that need updating like https://github.com/dashpay/platform/blob/feat/node-lts-v25/Dockerfile#L364

File: Dockerfile [hunk 4]
```diff
@@ -53,7 +53,7 @@ RUN apk add --no-cache \
         libc-dev \
         linux-headers \
         llvm-static llvm-dev  \
-        'nodejs~=16' \
+        nodejs \
```
Comment (shumkov) [inline Dockerfile:56]: Set specific version again?

File: packages/js-dapi-client/polyfills/fetch-polyfill.js (hunk 2 above)
Comment (shumkov) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: So we replace built-in `fetch` to `node-fetch`? What if we just do not do that?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dapi-client/polyfills/fetch-polyfill.js (hunk 2 above)
Comment (pshenmic) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: Js-dapi-client will be unable to connect with self-signed SSL hosts, because the code accepting such certificates is built on top of `http` library's agent, while built-in fetch is using `undici`. We can enable undici in the project and completely remove node-fetch, but its not yet ready for bundling in the web.

Comment (pshenmic) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: Yes you can, there is a webpack config for it already.

There is another approach tho, by using `cross-fetch` instead of `node-fetch`, but it uses factory approach to instantiate fetch, and a little code refactor will be required

File: Dockerfile [hunk 5]
```diff
@@ -361,7 +361,7 @@ RUN yarn workspaces focus --production @dashevo/dapi
 #
 # STAGE: FINAL DAPI IMAGE
 #
-FROM node:16-alpine3.16 AS dapi
+FROM node:18-alpine3.18 AS dapi
```
Comment (strophy) [inline Dockerfile:364]:
```suggestion
FROM node:18-alpine${ALPINE_VERSION} AS dapi
```

File: Dockerfile [hunk 6]
```diff
@@ -29,7 +29,7 @@
 # 3. Github Actions have shared networking configured, so we need to set a random
 # SCCACHE_SERVER_PORT port to avoid conflicts in case of parallel compilation
 
-ARG ALPINE_VERSION=3.16
+ARG ALPINE_VERSION=3.18
```
Comment (shumkov) [inline Dockerfile:32]: Why did you change the version of alpine?

File: packages/js-dapi-client/polyfills/fetch-polyfill.js (hunk 2 above)
Comment (shumkov) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: So undici doesn't accept http user agent?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 6 above)
Comment (pshenmic) [reply Dockerfile:32]: It was changed in the buildbase image

Comment (strophy) [reply Dockerfile:32]: This change is necessary because Alpine 3.16 only packages Node 16.

Comment (pshenmic) [reply Dockerfile:32]: According https://www.alpinelinux.org/posts/Alpine-3.18.0-released.html, it should ship node.js 20 (when installed with apk i believe), but in fact its node v18, here is output from dapi container:


```
/platform $ node -v
v18.17.1
/platform $ ```
> NAME="Alpine Linux"
> ID=alpine
> VERSION_ID=3.18.3
> PRETTY_NAME="Alpine Linux v3.18"
> HOME_URL="https://alpinelinux.org/"
> BUG_REPORT_URL="https://gitlab.alpinelinux.org/alpine/aports/-/issues"
> /platform $ node -v
> v18.17.1
```

Comment (pshenmic) [reply Dockerfile:32]: We also changed it here in the buildbase https://github.com/strophy/docker-buildbase/blob/master/Dockerfile (in the image 0.0.9)

Comment (shumkov) [reply Dockerfile:32]: oh ok, i just wondering if it works properly.

File: packages/js-dapi-client/polyfills/fetch-polyfill.js (hunk 2 above)
Comment (shumkov) [reply packages/js-dapi-client/polyfills/fetch-polyfill.js:3]: Discussed in Slack. It's only JSON RPC that is temporary thing and will be replaced with gRPC before MVP

**Review by shumkov — APPROVED** [review]
👍
